rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -0,0 +1,387 @@
+Instead, users can effectively stub local code (functions, methods, and types) using conditional compilation.
+For example, they could specify using `#[cfg(kani)]` to turn off the original definition and turn on the replacement definition when Kani is running, similarly to the ghost state approach taken in the [Tokio Bytes proof](https://model-checking.github.io/kani-verifier-blog/2022/08/17/using-the-kani-rust-verifier-on-tokio-bytes.html).
+
+The disadvantage with this approach is that it does not provide any way to stub external code, which is one of the main motivations of our proposed approach.
+
+### Alternative #2: Source-to-source transformation
+
+In this alternative, we rewrite the source code before it even gets to the compiler.
+The advantage with this approach is that it is very flexible, allowing us to stub functions, methods, and types, either by directly replacing them, or appending their replacements and injecting appropriate conditional compilation guards.
+
+This approach entails less user effort than Alternative #1, but it has the same downside that it requires all source code to be available.
+It also might be difficult to inject code in a way that names are correctly resolved (e.g., if the replacement code comes from a different crate).
+Also, source code is difficult to work with (e.g., unexpanded macros).
+
+On the last two points, we might be able to take advantage of an existing source analysis platform like `rust-analyzer` (which has facilities like structural search replace), but this would add more (potentially fragile) dependencies to Kani.
+
+### Alternative #3: AST-to-AST or HIR-to-HIR transformation
+
+In this alternative, we implement stubbing by rewriting the [AST](https://rustc-dev-guide.rust-lang.org/syntax-intro.html) or [High-Level IR (HIR)](https://rustc-dev-guide.rust-lang.org/hir.html) of the program.
+The HIR is a more compiler-friendly version of the AST; it is what is used for type checking.
+To swap out a function, method, or type at this level, it looks like it would be necessary to add another pass to `rustc` that takes the initial AST/HIR and produces a new AST/HIR with the appropriate replacements.
+
+The advantage with this approach is, like source transformations, it would be very flexible.
+The downside is that it would require modifying `rustc` (as far as we know, there is not an API for plugging in a new AST/HIR pass), and would also require performing the transformations at a very syntactic level: although the AST/HIR would likely be easier to work with than source code directly, it is still very close to the source code and not very abstract.
+Furthermore, it would require that we have access to the AST/HIR for all external code, and---provided we supported stubbing across crate boundaries---we would need to figure out how to inject the AST/HIR from one crate into another (the AST/HIR is usually just constructed for the crate currently being compiled).
+
+## Open questions
+
```

Comment:
How can the user verify that the stub is an abstraction of the original function/method?

Reply:
I'll add this - a good open question.